hydroflow_lang/src/graph/flat_graph_builder.rs

```diff
@@ -145,19 +223,109 @@ impl FlatGraphBuilder {
                     inn: rhs_ends.inn,
                 };
                 self.links.push(link_ends);
-                outer_ends
+                Ok(outer_ends)
             }
             Pipeline::Operator(operator) => {
                 let op_span = Some(operator.span());
                 let nid = self
                     .flat_graph
                     .insert_node(Node::Operator(operator), current_varname.cloned());
-                Ends {
+                Ok(Ends {
                     inn: Some((PortIndexValue::Elided(op_span), GraphDet::Determined(nid))),
                     out: Some((PortIndexValue::Elided(op_span), GraphDet::Determined(nid))),
+                })
+            }
+            Pipeline::Import(import) => {
+                // TODO: https://github.com/rust-lang/rfcs/pull/3200
+                // this would be way better...
+                let mut dir = self.macro_invocation_path.clone();
+                dir.pop();
+
+                let file_contents = std::fs::read_to_string(dir.join(import.filename.value()))
+                    .map_err(|e| {
+                        Error::new(
+                            import.filename.span(),
+                            format!("filename: {}, err: {e}", import.filename.value()),
+                        )
+                    })?;
+
+                // TODO: see also above, parse_str sets all the spans in the resulting parsed token stream to the parent macro invocation span.
+                // This means that any error inside the imported module will manifest as a giant red squiggly line under the parent hydroflow_syntax!{} call.
+                let statements = match syn::parse_str::<HfCode>(&file_contents) {
+                    Ok(code) => code,
+                    Err(err) => {
+                        self.diagnostics.push(err.clone().into());
+                        return syn::Result::Err(err);
+                    }
```

Comment:
Can we fix this todo with the following:

```
self.diagnostics.push(Diagnostic::spanned(import.span(), Level::Error, err.to_string());
```

Reply:
yeah that is much better, but still not ideal.